src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonPropertyInfo.cs

```diff
@@ -100,24 +99,95 @@ public abstract class JsonPropertyInfo
                 _shouldSerialize = value;
                 // By default we will go through faster path (not using delegate) and use IgnoreCondition
                 // If users sets it explicitly we always go through delegate
-                IgnoreCondition = null;
+                _ignoreCondition = null;
                 IsIgnored = false;
                 _shouldSerializeIsExplicitlySet = true;
             }
         }
 
-        private protected Func<object, object?, bool>? _shouldSerialize;
+        internal JsonIgnoreCondition? IgnoreCondition
+        {
+            get => _ignoreCondition;
+            set
+            {
+                Debug.Assert(!_isConfigured);
+
+                _ignoreCondition = value;
+                IsIgnored = value == JsonIgnoreCondition.Always;
+                _shouldSerialize = value != null ? GetShouldSerializeForIgnoreCondition(value.Value) : null;
+                _shouldSerializeIsExplicitlySet = false;
+            }
+        }
+
+        private Func<object, object?, bool>? _shouldSerialize;
         private bool _shouldSerializeIsExplicitlySet;
+        private JsonIgnoreCondition? _ignoreCondition;
+
+        private protected abstract Func<object, object?, bool> GetShouldSerializeForIgnoreCondition(JsonIgnoreCondition condition);
+
+        /// <summary>
+        /// Gets or sets a custom attribute provider for the current property.
+        /// </summary>
+        /// <remarks>
+        /// When resolving metadata via <see cref="DefaultJsonTypeInfoResolver"/> this
+        /// will be populated with the underlying <see cref="MemberInfo" /> of the serialized property or field.
+        ///
+        /// Setting a custom attribute provider will have no impact on the contract model,
```

Comment:
👍At one point we should also consider populating this for source gen but IMO optional for now since I'm not sure what effect this will have on linkability